### DISCUSSION TITLE
How to virtualize a WrapPanel with equal-sized elements ?

### DISCUSSION THREAD

**Discussion by kokaKOLA-NEW** [AvaloniaUI/Avalonia#15321] [Q&A] [answered]
Hello,

I am new to Avalonia and GitHub, and I am looking for a way to virtualize a WrapPanel with equal-sized elements. I have searched for solutions, but I couldn't find a clear answer on how to achieve this.

I came across a pull request on the AvaloniaUI GitHub repository (#11251) that adds a VirtualizingWrapPanel control, but it was closed and not merged into the master branch. I have a ListBox with a WrapPanel as the ItemsPanelTemplate, and I want to enable virtualization to improve performance when displaying a large number of items. However, I'm not sure how to implement virtualization with a WrapPanel in Avalonia.

Could someone please provide me with an example or point me in the right direction on how to virtualize a WrapPanel with equal-sized elements in Avalonia? Any help or guidance would be greatly appreciated.

Thank you!

**Comment by thevortexcloud** [comment]
There is nothing really stopping you from just copying the code in that PR into your project and changing the name spaces.

**Reply by thevortexcloud** [reply]
Oh also, you can use ItemsRepeater with a uniform grid. But it can be fairly complicated depending on what you are doing. If you don't need selection it's not too bad. The Avalonia items repeater is based on the UWP one, so the UWP docs/examples are just as applicable to Avalonia.

https://learn.microsoft.com/en-us/windows/apps/design/controls/items-repeater

**Reply by kokaKOLA-NEW** [reply]
I took a look at ItemsRepeater, but yes, it does seem quite complex.

**Comment by kokaKOLA-NEW** [comment]
😅 Sorry, I didn't realize I could view the code directly in the PR. Do I need to ask emmauss for permission before using this code in my project?

**Reply by thevortexcloud** [reply]
Likely not since it was contributed to Avalonia under MIT, which lets you do pretty much anything you want. But if you want to be safe, feel free to ask. They are active around here.

**Reply by emmauss** [reply]
You are free to use the code.

**Reply by kokaKOLA-NEW** [reply]
Thank you

**Comment by kokaKOLA-NEW** [comment]
Is the addition of the VirtualizingUniformGrid control planned for a future version of the project, as proposed in the PR discussion at https://github.com/AvaloniaUI/Avalonia/pull/11251 ? This would be useful since the vast majority of wrap panels use items of equal size.

**Reply by thevortexcloud** [reply]
ItemsRepeater already has UniformGrid support.

**Reply by kokaKOLA-NEW** [reply, marked as answer]
Ha yes, 😅 I just had to do that :

```xaml
<ItemsRepeater>
          <ItemsRepeater.Layout>
                    <UniformGridLayout MinColumnSpacing="10" MinRowSpacing="10" Orientation="Horizontal" />
          </ItemsRepeater.Layout>
          <ItemsRepeater.ItemTemplate>
                    <DataTemplate>
                            <!-- ... -->                       
                    </DataTemplate>
          </ItemsRepeater.ItemTemplate>
</ItemsRepeater>```

Thank you anyway.
